lightningd/subd.c

```diff
@@ -19,53 +19,72 @@
 #include <wire/common_wiregen.h>
 #include <wire/wire_io.h>
 
-static bool move_fd(int from, int to)
+/* Carefully move fd *@from to @to: on success *from set to to */
+static bool move_fd(int *from, int to)
 {
-	assert(from >= 0);
+	assert(*from >= 0);
 
 	/* dup2 with same arguments may be a no-op, but
 	 * the later close would make the fd invalid.
 	 * Handle this edge case.
 	 */
-	if (from == to)
+	if (*from == to)
 		return true;
 
-	if (dup2(from, to) == -1)
+	if (dup2(*from, to) == -1)
 		return false;
 
 	/* dup2 does not duplicate flags, copy it here.
 	 * This should be benign; the only POSIX-defined
 	 * flag is FD_CLOEXEC, and we only use it rarely.
 	 */
-	if (fcntl(to, F_SETFD, fcntl(from, F_GETFD)) < 0)
+	if (fcntl(to, F_SETFD, fcntl(*from, F_GETFD)) < 0)
 		return false;
 
-	close(from);
+	close(*from);
+	*from = to;
 	return true;
 }
 
-/* Like the above, but move the fd from whatever it currently has
- * to any other unused fd number that is *not* its current value.
- */
-static bool move_fd_any(int *fd)
+/* Returns index of fds which is == this fd, or -1 */
+static int fd_used(const int *fds, size_t num_fds, int fd)
 {
-	int old_fd = *fd;
-	int new_fd;
-
-	assert(old_fd >= 0);
-
-	if ((new_fd = dup(old_fd)) == -1)
-		return false;
-
-	/* dup does not duplicate flags.  */
-	if (fcntl(new_fd, F_SETFD, fcntl(old_fd, F_GETFD)) < 0)
-		return false;
-
-	close(old_fd);
+	for (size_t i = 0; i < num_fds; i++) {
+		if (fds[i] == fd)
+			return i;
+	}
+	return -1;
+}
 
-	*fd = new_fd;
+/* Move an series of fd pointers into 0, 1, ... */
+static bool shuffle_fds(int *fds, size_t num_fds)
+{
+	/* If we need to move an fd out the way, this is a good place to start
+	 * looking */
+	size_t next_free_fd = num_fds;
+	for (size_t i = 0; i < num_fds; i++) {
+		int in_the_way;
+
+		/* Already in the right place?  Great! */
+		if (fds[i] == i)
+			continue;
+		/* Is something we care about in the way? */
+		in_the_way = fd_used(fds + i, num_fds - i, i);
+		if (in_the_way != -1) {
+			/* Find a high-numbered unused fd. */
+			while (fd_used(fds + i, num_fds - i, next_free_fd) != -1)
```

Comment:
`fd_used` is O(n), and this `while` loop potentially loops up to n times as well, and all this is in a loop that is called n times, too.

How about just using `dup`?  Presumably the OS has a fast O(1) or O(log n) structure for looking up file descriptor numbers.  i.e. retain `move_fd_any` and do something like:

```c
if (in_the_way != -1)
    /* Trick: in_the_way is offset by i! */
    move_fd_any(&fds[i + in_the_way]);
```

This *could* move the fd to another low index, but we will do another check on the next loop iteration anyway.  Assuming the OS has an O(1) structure for `dup`, then this drops the loop down to O( n^2 ) instead of O( n^3 ).  Though that may not matter overmuch since we do not expect to pass a lot of `fd`s around...

Reply:
N is so low here that I don't think it's worth *any* optimization.  Simplest wins.